### ISSUE TITLE
More syntax for creating intervals

### ISSUE THREAD

**Issue by dpsanders** [dpsanders/ValidatedNumerics.jl#76]
Allow the following for creating intervals:
- [x] `I"[1, 2]"`
- [ ] `II[3, 4]`
- [x] `1..2`

- ~~`3 + "[2,3]"`~~   (no reference to intervals)
- ~~`@interval [3, 4]`~~   (correctly creates a vector of intervals)

**Comment by dpsanders** [comment]
The second is easiest:

```julia
macro I_str(ex)
    @interval(ex)
end

julia> I"3"
[3.0, 3.0]

julia> I"[3, 4]"
[3.0, 4.0]

```

**Comment by dpsanders** [comment]
The third parses as follows
```julia
julia> dump(parse("II[3, 4]"))
Expr 
  head: Symbol ref
  args: Array(Any,(3,))
    1: Symbol II
    2: Int64 3
    3: Int64 4
  typ: Any
```

**Comment by dpsanders** [comment]
The third can be implemented as follows:
```julia
julia> immutable II_type end

julia> II = II_type()
II_type()

julia> using ValidatedNumerics

julia> Base.getindex(x::II_type, a, b) = @interval(a, b)
getindex (generic function with 120 methods)

julia> II[3,4]
[3.0, 4.0]

julia> Base.getindex(x::II_type, a) = @interval(a)
getindex (generic function with 121 methods)

julia> II[3.1]
[3.0999999999999996, 3.1]
```

**Comment by lbenet** [comment]
The problem you spotted with the second option is related to the following:
```julia
julia> @interval("[0,1]")  # works fine!
[0.0, 1.0]

julia> @interval("[0.1]") # throws an error!!
ERROR: ArgumentError: Unable to process string [0.1] as interval
 [inlined code] from essentials.jl:61
 in split_interval_string at /Users/benet/.julia/v0.5/ValidatedNumerics/src/intervals/rounding.jl:70
 in make_interval at /Users/benet/.julia/v0.5/ValidatedNumerics/src/intervals/rounding.jl:104
```
That is, `@interval` does not work with a string in the form of an interval, but with one single entry. Actually, I think that's ok.

**Comment by lbenet** [comment]
From the options above, at most I like the second one, because it is *julian*.

**Comment by lbenet** [comment]
Incidentally, the first one yields the following:
```julia
julia> @interval [1, 2]
2-element Array{ValidatedNumerics.Interval{Float64},1}:
 [1.0, 1.0]
 [2.0, 2.0]
```

So `transform` works, and creates a vector of arrays. That could be ok; yet, the following is a bit awkward:
```julia
julia> @interval [[1,2]; [2,3]]
4-element Array{ValidatedNumerics.Interval{Float64},1}:
 [1.0, 1.0]
 [2.0, 2.0]
 [2.0, 2.0]
 [3.0, 3.0]
```
(I used `;` to concatenate, to avoid a warning message.)

**Comment by lbenet** [comment]
In my opinion, since we are interfering a bit with the usual vector notation, we should avoid things like `1+"[2,3]"`

**Comment by dpsanders** [comment]
Yes, I think we should remove the last one.

The behaviour you notice is expected, since that is how `;` works:
julia> [[1,2]; [3,4]]
4-element Array{Int64,1}:
 1
 2
 3
 4

julia> @interval [[1,2]; [3,4]]
4-element Array{ValidatedNumerics.Interval{Float64},1}:
 [1.0, 1.0]
 [2.0, 2.0]
 [3.0, 3.0]
 [4.0, 4.0]
```

However, the following is, in some sense, nice:
```julia
julia> @interval [1 2; 3 4]
2x2 Array{ValidatedNumerics.Interval{Float64},2}:
 [1.0, 1.0]  [2.0, 2.0]
 [3.0, 3.0]  [4.0, 4.0]
```
So maybe we should not try to do anything special for `@interval [1,2]` and leave it as it is.

**Comment by dpsanders** [comment]
(And regard it as a feature, not a bug ;) )

**Comment by lbenet** [comment]
I agree that the behavior is expected, a feature ;-). 
Yet, if I wanted to construct the `Interval` vector with entries `[1,2]` and `[3,4]`, and clearly that was not the way to get it.

My point is that we should avoid any notation that is ambiguous (or may yield ambiguities) with the usual vector notation. We should only accept that kind of things, at most, when they appear as strings.

**Comment by dpsanders** [comment]
To do that in Julia, you currently have to do 
```
Vector{Int}[[1, 2], [3, 4]]
```
In v0.5 this should be changed to
```
[[1, 2], [3, 4]]
```
at which point you will be able to do
```
@interval [[1, 2], [3, 4]]
```
Currently this is not possible, since in Julia you must write
```
Vector{Int}[[1,2], [3,4]]
```

**Comment by dpsanders** [comment]
We should document that you can use `@interval` with a vector and matrix like this.

**Comment by lbenet** [comment]
I agree

**Comment by dpsanders** [comment]
Implemented in #88
